go/ct/rlz/conditions_test.go

```diff
@@ -426,3 +427,124 @@ func TestCondition_InOutOfRangeGetTestValues(t *testing.T) {
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			for i := 0; i < 1000; i++ {
+				gen := gen.NewStateGenerator()
+				test.condition.Restrict(gen)
+				state, err := gen.Generate(rnd)
+				if err != nil {
+					t.Fatalf("failed to build state: %v", err)
+				}
+				if checked, err := test.condition.Check(state); err != nil || !checked {
+					t.Errorf("failed to check condition: %v", err)
+				}
+			}
+		})
+	}
+}
+
+func TestCondition_BlobHashes_check(t *testing.T) {
+	tests := map[string]struct {
+		condition Condition
+		setup     func(*st.State)
+	}{
+		"hasBlobHash": {
+			condition: HasBlobHash(Param(0)),
+			setup: func(state *st.State) {
+				state.TransactionContext.BlobHashes = []vm.Hash{{0}}
+				state.Stack.Push(NewU256(0))
+			},
+		},
+		"hasNotBlobHash": {
+			condition: HasNoBlobHash(Param(0)),
+			setup: func(state *st.State) {
+				state.TransactionContext.BlobHashes = []vm.Hash{{0}}
+				state.Stack.Push(NewU256(1))
+			},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			for i := 0; i < 1000; i++ {
+				state := st.NewState(st.NewCode([]byte{}))
+				test.setup(state)
+
+				if checked, err := test.condition.Check(state); err != nil || !checked {
+					t.Errorf("failed to check condition: %v", err)
+				}
+			}
+		})
+	}
+}
+
+func TestCondition_BlobHashesProducesUnsatisfiable(t *testing.T) {
+	tests := map[string]struct {
+		condition Condition
+		setup     func(*st.State)
+	}{
+		"hasBlobHash-first": {
+			condition: And(HasBlobHash(Param(0)), HasNoBlobHash(Param(0))),
+			setup: func(state *st.State) {
+				state.Stack.Push(NewU256(0))
+			},
+		},
+		"hasNotBlobHash-first": {
+			condition: And(HasNoBlobHash(Param(0)), HasBlobHash(Param(0))),
+			setup: func(state *st.State) {
+				state.Stack.Push(NewU256(0))
+			},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			state := st.NewState(st.NewCode([]byte{}))
+			test.setup(state)
+
+			gen := gen.NewStateGenerator()
+			test.condition.Restrict(gen)
+			_, err := gen.Generate(rand.New(0))
+			if err == nil {
+				t.Errorf("expected unsatisfiable condition")
+			}
+		})
+	}
+}
+
+func TestCondition_BlobHashesPRoducesGetTestValues(t *testing.T) {
+	tests := map[string]Condition{
+		"hasBlobHash":   HasBlobHash(Param(0)),
+		"hasNoBlobHash": HasNoBlobHash(Param(0)),
+	}
+
+	for name, condition := range tests {
+		t.Run(name, func(t *testing.T) {
+			testValues := condition.GetTestValues()
+			if len(testValues) != 2 {
+				t.Fatalf("unexpected amount test values, got %v, wanted 2", len(testValues))
+			}
+
+			if testValues[0].(*testValue[bool]).value != true {
+				t.Errorf("unexpected test value, got %v, wanted true ", testValues[0].(*testValue[bool]).value)
+			}
+
+			if testValues[1].(*testValue[bool]).value != false {
+				t.Errorf("unexpected test value, got %v, wanted false ", testValues[1].(*testValue[bool]).value)
+			}
+
```

Comment:
It is generally better to test for properties of the expected result, not implementation details. For instance, here, the implementation details like the number of generated test cases or their type should ideally not be exploited.

A better way to test this could look like this:
```suggestion
			var matchCovered, failCovered bool
			for _, value := range condition.GetTestValues() {
				gen := gen.NewStateGenerator()
				value.Restrict(gen)
				state, err := gen.Generate(rand.New(0))
				if err != nil {
					t.Fatalf("failed to build state: %v", err)
				}
				if matches, err := condition.Check(state); err != nil {
					t.Errorf("failed to check condition: %v", err)
				} else if matches {
					matchCovered = true
				} else {
					failCovered = true
				}
			}

			if !matchCovered {
				t.Errorf("no test value matched the condition")
			}

			if !failCovered {
				t.Errorf("no test value failed the condition")
			}
```
This way, the intention is tested and only public interfaces are utilized.